src/stream/src/executor/watermark_filter.rs

```diff
@@ -221,38 +232,29 @@ impl<S: StateStore> WatermarkFilterExecutor<S> {
         table: &StateTable<S>,
         watermark_type: DataType,
     ) -> StreamExecutorResult<Datum> {
-        let watermark_iter_futures = (0..VIRTUAL_NODE_COUNT).map(|vnode| {
-            async move {
-                let pk = Row::new(vec![Some(ScalarImpl::Int16(vnode as _))]);
-                let watermark_iter = table.iter_with_pk_prefix(&pk).await?.fuse();
-                pin_mut!(watermark_iter);
-                let watermark_row = watermark_iter.next().await;
-
-                // The remote storage should only contains at most 1 watermark for each vnode.
-                let next_row = watermark_iter.next().await;
-                if next_row.is_some() {
-                    bail!("Should not have more than on value for a given vnode")
-                }
-
-                match watermark_row {
-                    Some(row) => {
-                        let row = row?;
-                        if row.size() == 1 {
-                            Ok::<_, StreamExecutorError>(row[0].to_owned())
-                        } else {
-                            bail!("The watermark row should only contains 1 datum");
-                        }
+        let watermark_iter_futures = (0..VIRTUAL_NODE_COUNT).map(|vnode| async move {
+            let pk = Row::new(vec![Some(ScalarImpl::Int16(vnode as _))]);
+            let watermark_row = table.get_row(&pk).await?;
+            match watermark_row {
+                Some(row) => {
+                    if row.size() == 1 {
+                        Ok::<_, StreamExecutorError>(row[0].to_owned())
+                    } else {
+                        bail!("The watermark row should only contains 1 datum");
                     }
-                    _ => Ok(None),
                 }
+                _ => Ok(None),
             }
         });
         let watermarks: Vec<_> = join_all(watermark_iter_futures)
             .await
             .into_iter()
             .try_collect()?;
+
         let watermark = watermarks.into_iter().max().unwrap();
```

Comment:
I think we can assert all `None` or all `Some` here, and avoid relying on the derived `Ord` for `Datum = Option<_>` here. :)